File: src/FluentCards/InputNumberBuilder.cs
namespace FluentCards;

/// <summary>
/// Fluent builder for creating InputNumber elements.
/// </summary>
public class InputNumberBuilder
{
    private readonly InputNumber _input = new();

    /// <summary>
    /// Sets the unique identifier for the input.
    /// </summary>
    /// <param name="id">The unique identifier.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputNumberBuilder WithId(string id)
    {
        _input.Id = id;
        return this;
    }

    /// <summary>
    /// Sets the label for the input.
    /// </summary>
    /// <param name="label">The input label.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputNumberBuilder WithLabel(string label)
    {
        _input.Label = label;
        return this;
    }

    /// <summary>
    /// Sets the placeholder text for the input.
    /// </summary>
    /// <param name="placeholder">The placeholder text.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputNumberBuilder WithPlaceholder(string placeholder)
    {
        _input.Placeholder = placeholder;
        return this;
    }

    /// <summary>
    /// Sets the default value for the input.
    /// </summary>
    /// <param name="value">The default value.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputNumberBuilder WithValue(double value)
    {
        _input.Value = value;
        return this;
    }

    /// <summary>
    /// Sets the minimum value allowed.
    /// </summary>
    /// <param name="min">The minimum value.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputNumberBuilder WithMin(double min)
    {
        _input.Min = min;
        return this;
    }

    /// <summary>
    /// Sets the maximum value allowed.
    /// </summary>
    /// <param name="max">The maximum value.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputNumberBuilder WithMax(double max)
    {
        _input.Max = max;
        return this;
    }

    /// <summary>
    /// Sets whether the input is required.
    /// </summary>
    /// <param name="required">True if required, false otherwise.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputNumberBuilder IsRequired(bool required = true)
    {
        _input.IsRequired = required;
        return this;
    }

    /// <summary>
    /// Sets the error message to display when validation fails.
    /// </summary>
    /// <param name="message">The error message.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputNumberBuilder WithErrorMessage(string message)
    {
        _input.ErrorMessage = message;
        return this;
    }

    /// <summary>
    /// Builds and returns the configured InputNumber.
    /// </summary>
    /// <returns>The configured InputNumber instance.</returns>
    public InputNumber Build()
    {
        return _input;
    }
}

File: go/fluentcards/input_number_builder.go
package fluentcards

// InputNumberBuilder builds an Input.Number Adaptive Card element.
type InputNumberBuilder struct {
	data map[string]any
}

func newInputNumberBuilder() *InputNumberBuilder {
	return &InputNumberBuilder{data: map[string]any{"type": "Input.Number", "id": ""}}
}

func (b *InputNumberBuilder) WithID(id string) *InputNumberBuilder {
	b.data["id"] = id
	return b
}

func (b *InputNumberBuilder) WithLabel(label string) *InputNumberBuilder {
	b.data["label"] = label
	return b
}

func (b *InputNumberBuilder) WithPlaceholder(placeholder string) *InputNumberBuilder {
	b.data["placeholder"] = placeholder
	return b
}

func (b *InputNumberBuilder) WithValue(value float64) *InputNumberBuilder {
	b.data["value"] = value
	return b
}

func (b *InputNumberBuilder) WithMin(min float64) *InputNumberBuilder {
	b.data["min"] = min
	return b
}

func (b *InputNumberBuilder) WithMax(max float64) *InputNumberBuilder {
	b.data["max"] = max
	return b
}

func (b *InputNumberBuilder) WithIsRequired(isRequired bool) *InputNumberBuilder {
	b.data["isRequired"] = isRequired
	return b
}

func (b *InputNumberBuilder) WithErrorMessage(errorMessage string) *InputNumberBuilder {
	b.data["errorMessage"] = errorMessage
	return b
}

func (b *InputNumberBuilder) WithSpacing(spacing Spacing) *InputNumberBuilder {
	b.data["spacing"] = string(spacing)
	return b
}

func (b *InputNumberBuilder) WithIsVisible(isVisible bool) *InputNumberBuilder {
	b.data["isVisible"] = isVisible
	return b
}

func (b *InputNumberBuilder) WithSeparator(separator bool) *InputNumberBuilder {
	b.data["separator"] = separator
	return b
}

func (b *InputNumberBuilder) WithHeight(height string) *InputNumberBuilder {
	b.data["height"] = height
	return b
}

func (b *InputNumberBuilder) WithFallback(fallback any) *InputNumberBuilder {
	b.data["fallback"] = fallback
	return b
}

func (b *InputNumberBuilder) WithRequires(key, version string) *InputNumberBuilder {
	reqs, ok := b.data["requires"].(map[string]any)
	if !ok {
		reqs = map[string]any{}
	}
	reqs[key] = version
	b.data["requires"] = reqs
	return b
}

func (b *InputNumberBuilder) WithRtl(rtl bool) *InputNumberBuilder {
	b.data["rtl"] = rtl
	return b
}

func (b *InputNumberBuilder) WithLabelPosition(position InputLabelPosition) *InputNumberBuilder {
	b.data["labelPosition"] = string(position)
	return b
}

func (b *InputNumberBuilder) WithLabelWidth(width string) *InputNumberBuilder {
	b.data["labelWidth"] = width
	return b
}

func (b *InputNumberBuilder) WithInputStyle(style InputStyle) *InputNumberBuilder {
	b.data["inputStyle"] = string(style)
	return b
}

func (b *InputNumberBuilder) Build()Card {
	return b.data
}
